client/consensus/babe/rpc/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+	StringError(String)
+}
+
+impl From<Error> for jsonrpc_core::Error {
+	fn from(error: Error) -> Self {
+		jsonrpc_core::Error {
+			message: format!("{}", error).into(),
+			code: jsonrpc_core::ErrorCode::ServerError(1234),
+			data: None,
+		}
+	}
+}
+
+/// fetches the epoch data for a given slot_number.
+fn epoch_data<B, C, SC>(
+	epoch_changes: &SharedEpochChanges<B, Epoch>,
+	client: &Arc<C>,
+	babe_config: &Config,
+	slot_number: u64,
+	select_chain: &SC,
+) -> Result<Epoch, Error>
+	where
+		B: BlockT,
+		C: HeaderBackend<B> + HeaderMetadata<B, Error=BlockChainError> + 'static,
+		SC: SelectChain<B>,
+{
+	let parent = select_chain.best_chain()?;
+	epoch_changes.lock().epoch_for_child_of(
+		descendent_query(&**client),
+		&parent.hash(),
+		parent.number().clone(),
+		slot_number,
+		|slot| babe_config.genesis_epoch(slot),
+	)
+		.map_err(|e| Error::Consensus(ConsensusError::ChainLookup(format!("{:?}", e))))?
+		.map(|e| e.into_inner())
+		.ok_or(Error::Consensus(ConsensusError::InvalidAuthoritiesSet))
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use substrate_test_runtime_client::{
+		DefaultTestClientBuilderExt,
+		TestClientBuilderExt,
+		TestClientBuilder,
+	};
+	use sp_application_crypto::AppPair;
+	use sp_keyring::Ed25519Keyring;
+	use sc_keystore::Store;
+
+	use std::sync::Arc;
+	use sc_consensus_babe::{Config, block_import, AuthorityPair};
+	use jsonrpc_core::IoHandler;
+
+	/// creates keystore backed by a temp file
+	fn create_temp_keystore<P: AppPair>(authority: Ed25519Keyring) -> (KeyStorePtr, tempfile::TempDir) {
+		let keystore_path = tempfile::tempdir().expect("Creates keystore path");
+		let keystore = Store::open(keystore_path.path(), None).expect("Creates keystore");
+		keystore.write().insert_ephemeral_from_seed::<P>(&authority.to_seed())
+			.expect("Creates authority key");
+
+		(keystore, keystore_path)
+	}
+
+	#[test]
+	fn rpc() {
+		let builder = TestClientBuilder::new();
+		let (client, longest_chain) = builder.build_with_longest_chain();
+		let client = Arc::new(client);
+		let config = Config::get_or_compute(&*client).expect("config available");
+		let (_, link) = block_import(
+			config.clone(),
+			client.clone(),
+			client.clone(),
+			client.clone(),
+		).expect("can initialize block-import");
+
+		let epoch_changes = link.epoch_changes().clone();
+		let select_chain = longest_chain;
+		let keystore = create_temp_keystore::<AuthorityPair>(Ed25519Keyring::Alice).0;
+		let handler = BabeRPCHandler::new(client.clone(), epoch_changes, keystore, config, select_chain);
+		let mut io = IoHandler::new();
+
+		io.extend_with(BabeApi::to_delegate(handler));
+		let request = r#"{"jsonrpc":"2.0","method":"babe_epochAuthorship","params": [],"id":1}"#;
+		let response = r#"{"jsonrpc":"2.0","result":{"5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY":{"primary":[0],"secondary":[1,2,4]}},"id":1}"#;
+
+		assert_eq!(Some(response.into()), io.handle_request_sync(request));
+	}
+}
```

Comment:
```suggestion
}
```

newline at end of file.